pyiceberg/table/__init__.py

```diff
@@ -1198,10 +1198,11 @@ def upsert(
 
                 update_row_cnt = len(rows_to_update)
 
-                # build the match predicate filter
-                overwrite_mask_predicate = upsert_util.create_match_filter(rows_to_update, join_cols)
+                if len(rows_to_update) > 0:
+                    # build the match predicate filter
+                    overwrite_mask_predicate = upsert_util.create_match_filter(rows_to_update, join_cols)
 
-                tx.overwrite(rows_to_update, overwrite_filter=overwrite_mask_predicate)
+                    tx.overwrite(rows_to_update, overwrite_filter=overwrite_mask_predicate)
```

Comment:
should we enforce not producing empty snapshots lower in the stack, in the write functions themselves (`overwrite`/`append`)?

Reply:
Yes, that's a great suggestion. I was also looking into that, but there are two caveats:

- Before computing all the manifests, we don't know if they are empty. The delete manifests are lazily computed, so making sure that we short-circuit earlier in the process, makes sure that we don't do all the work.
- There is also a behavioral aspect around it, for example, when you do `tbl.append(df)` where `df` is empty, or `tbl.delete('x = 10')`, you may still want to produce a snapshot to indicate that something happened. My opinion is that it should result in a no-op, but Java creates a snapshot:

![image](https://github.com/user-attachments/assets/2a3c3c81-f80d-4c59-9614-852951360a0c)

I believe @sungwy and I had a similar discussion.